### PR TITLE
CONTRIBUTING: loose squash merging rules

### PR DESCRIPTION
**Pull request by SuperSandro2000**
This mainly puts the burden to not squash merge on the PR author and keeps it as a loophole if the author does not keep the PR in accordance to this guide.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: CONTRIBUTING.md
Comment (ncfavier) [inline CONTRIBUTING.md:58]:
```suggestion
Pull requests should not be squash merged in order to keep complete commit messages and GPG signatures intact and must not when the change doesn't make sense as a single commit.
This means that, when addressing review comments in order to keep the pull request in an always mergeable status, you will sometimes need to rewrite your branch's history and then force-push it with `git push --force-with-lease`.
Useful git commands that can help a lot with this are `git commit --patch --amend` and `git rebase --interactive @~3`. For more details consult the git man pages.
```

**Review by ncfavier — APPROVED** [review]
Please squash 😅

File: CONTRIBUTING.md
Comment (primeos) [inline CONTRIBUTING.md:58]: FYI: I would've omitted the `@~3` from `git rebase --interactive` as that causes even more trouble if users execute that command without understanding it (e.g., they will duplicate two commits if they execute it on their PR branch that only contains a single additional commit - the main problem being that some Nixpkgs maintainers might not even notice it and we'll have duplicate commits in the history after the merge).

(Using something like `git merge-base --fork-point origin/master` could be a bit better but also has lot's of issues as it too makes lots of assumptions (`origin/master`) so it doesn't really make sense either.)

It would probably be much better to link to some external documentation like https://git-rebase.io/ or https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History instead.

PS: You discarded the commit message body when you used GHs squash merge functionality to merge this PR (which is both ironic and undesirable as https://github.com/NixOS/nixpkgs/commit/756a5e6b74f786f08fa3a4c006b31fa316133566 now lacks any reasoning for the change... :o).

PPS: +1 for the `complete commit messages and GPG signatures` part

Comment (ncfavier) [reply CONTRIBUTING.md:58]: I like the idea of omitting `@~3` and linking each git command to its manual (or third-party documentation).

Comment (SuperSandro2000) [reply CONTRIBUTING.md:58]: #207487

**Pull request merged**
